SDK wait provider to be ready (#80)
open-feature/ofep: OFEP-sdk-wait-provider-ready.md

The `initialize` function is called asynchronously and it is possible to get the information that the provider is ready while awaiting an initial ready event.

In some languages such as `javascript` waiting for the event could be one line of code, but on other languages it could need way more boiler plate code to do that.
```javascript
Openfeature.setProvider(my-provider)
const client = Openfeature.getClient()
await new Promise((resolve) => client.on(‘ready’, resolve))
```

```java
OpenFeatureAPI.getInstance().setProvider("test", g);
Client cli = OpenFeatureAPI.getInstance().getClient("test");
CompletableFuture<EventDetails> completableFuture = new CompletableFuture<>();
OpenFeatureAPI.getInstance().onProviderReady(new Consumer<EventDetails>() {
    @Override
    public void accept(EventDetails eventDetails) {
        completableFuture.complete(eventDetails);
    }
});
completableFuture.get();
```

## Proposal
It would be great to have a way in the SDKs to wait for the provider to be ready.

## Alternatives

### Alternative 1: make `setProvider` synchronous

> Probably not the most ideal solution but we could change the behavior of `setProvider` to wait for the initialize function to throw or return.  

Reviewer: Depending on the language async is not always that great for waiting for something that you want to actually allow to complete asynchronously. (Java before completable futures, Rust where generally it will only progress if you are polling it.) Likely you end up making two methods anyway.

*It is a breaking change from the actual implementation.*

> ### Alternative 2: Add a waiting function in the SDKs to wait for the provider to be ready

Reviewer: I think this is reasonable. It provides a good deal of flexibility. I think, depending on language, you still end up also adding a timeout. Waiting forever isn't a great idea, and similar to the initial problem not all languages provide the ease of handling the timeout. (In JS you just race the promise with a time promise.)

`waitReady(timeout)`.

I do wonder if there is any value in difference here between core spec and stuff you can already build with the spec? It is inconvenient in java, but not impossible. Is there some other level we should have for things that can be implemented as a layer? I suppose if it is just an optional part of the spec that would be sufficient.

Reviewer 2: > I think this is reasonable. It provides a good deal of flexibility. I think, depending on language, you still end up also adding a timeout. Waiting forever isn't a great idea, and similar to the initial problem not all languages provide the ease of handling the timeout. (In JS you just race the promise with a time promise.)

Agreed 100%.

> I do wonder if there is any value in difference here between core spec and stuff you can already build with the spec? It is inconvenient in java, but not impossible. Is there some other level we should have for things that can be implemented as a layer? I suppose if it is just an optional part of the spec that would be sufficient.

I've been considering something like this as well, but I think an optional requirement (MAY or SHOULD) would be sufficient for this.

MAY or SHOULD, in combination with the nature of the requirement, strongly imply this would be a convenience method, which I believe is basically the case here.

Adding a waiting function will allow to block the SDK until the provider is ready.  
This new function will wait for a `ready` event or a timelimit for the provider to be ready.  

```javascript
Openfeature.setProvider(myprovider)
const client = Openfeature.getClient()
await client.isReady()
```

> ### Alternative 2: Chain `waitReady()` with `setProvider(...)`

Reviewer 3: I like that you can wait for a provider to be ready before a client has even been created. This helps keep a nice separation between the [application integrator](https://openfeature.dev/specification/glossary#application-integrator) and the [application author](https://openfeature.dev/specification/glossary#application-author).

Author: Yes, I like the fact that we are waiting for the `provider` to be ready and not the `client` to be ready.

> ### Alternative 2: Chain `waitReady()` with `setProvider(...)`

Reviewer 4: ```suggestion
### Alternative 3: Chain `waitReady()` with `setProvider(...)`
```

> ### Alternative 2: Chain `waitReady()` with `setProvider(...)`

Reviewer 4: Chaining essentially require to change the signature (hence a breaking change) and adds complexity for SDK implementation.

I would rather expose a dedicated blocking _setProvider_ along with current async _setpProvider_ . For example, Java gRPC expose similar methods - `RouteGuideGrpc.newBlockingStub` vs `RouteGuideGrpc.newStub` [^1] 

[^1]: https://grpc.io/docs/languages/java/basics/#instantiating-a-stub

Author: Should we add the option of using a blocking `setProvider` as an alternative in the OFEP?
How should we call this function? `setProviderSync`, `setProviderBlocking`, `setProviderAndWait` ?

Reviewer 3: Yeah, I think an alternative 4 would make sense. The name itself isn't that important at this stage of the proposal.

Reviewer 4: +1 for alternative 4

Add a chain function to `setProvider(...)` to wait until the intialization is done.

> It will wait until the `initialize` function throw or return.

Reviewer 3: ```suggestion
It will wait for the `initialize` function to throw or return.
```

```javascript

> Openfeature.setProvider(myprovider).waitReady()

Reviewer 3: ```suggestion
await Openfeature.setProvider(myprovider).waitReady()
```

I'm assuming, in this case, the `waitReady` function would return a promise.

## Background

Provider can now define a `initialize` function called by the SDK when Openfeature users call the function `setProvider(...)`. 
The `initialize` function is called asynchronously and it is possible to get the information that the provider is ready while awaiting an initial ready event.

In some languages such as `javascript` waiting for the event could be one line of code, but on other languages it could need way more boiler plate code to do that.
```javascript
Openfeature.setProvider(my-provider)
const client = Openfeature.getClient()
await new Promise((resolve) => client.on(‘ready’, resolve))
```

```java
OpenFeatureAPI.getInstance().setProvider("test", g);
Client cli = OpenFeatureAPI.getInstance().getClient("test");
CompletableFuture<EventDetails> completableFuture = new CompletableFuture<>();
OpenFeatureAPI.getInstance().onProviderReady(new Consumer<EventDetails>() {
    @Override
    public void accept(EventDetails eventDetails) {
        completableFuture.complete(eventDetails);
    }
});
completableFuture.get();
```

## Proposal
It would be great to have a way in the SDKs to wait for the provider to be ready.

> 

Reviewer 3: @thomaspoignant, could you please update the proposal section with an example from section 4? I think we could approve and merge after that.

Author: I've added examples + change the status of the OFEP.

## Alternatives

### Alternative 1: make `setProvider` synchronous
Probably not the most ideal solution but we could change the behavior of `setProvider` to wait for the initialize function to throw or return.  
*It is a breaking change from the actual implementation.*

### Alternative 2: Add a waiting function in the SDKs to wait for the provider to be ready
Adding a waiting function will allow to block the SDK until the provider is ready.  
This new function will wait for a `ready` event or a timelimit for the provider to be ready.  

```javascript
Openfeature.setProvider(myprovider)
const client = Openfeature.getClient()
await client.isReady()
```

### Alternative 3: Chain `waitReady()` with `setProvider(...)`
Add a chain function to `setProvider(...)` to wait until the intialization is done.
It will wait for the `initialize` function to throw or return.

```javascript
await Openfeature.setProvider(myprovider).waitReady()
const client = Openfeature.getClient()
```

> ### Alternative 4: Add a blocking `setProvider`

Reviewer 3: This is my preferred option.

Author: Same for me, I would go for option 4.

Reviewer 4: Given that we have existing reference implementations on other frameworks (ex:- gRPC), I am also voting for this option. I think this is the easiest option with respect to implementation efforts.
